src/lib/coda_base/pending_coinbase.ml

```diff
@@ -0,0 +1,499 @@
+
+  module Checked = struct
+    open Coinbase_stack
+
+    type var = Hash.var
+
+    module Merkle_tree =
+      Snarky.Merkle_tree.Checked
+        (Tick)
+        (struct
+          type value = Pedersen.Checked.Digest.t
+
+          type var = Pedersen.Checked.Digest.var
+
+          let typ = Pedersen.Checked.Digest.typ
+
+          let hash ~height h1 h2 =
+            let to_triples (bs : Pedersen.Checked.Digest.Unpacked.var) =
+              Bitstring_lib.Bitstring.pad_to_triple_list
+                ~default:Boolean.false_
+                (bs :> Boolean.var list)
+            in
+            let%bind h1 = Pedersen.Checked.Digest.choose_preimage h1
+            and h2 = Pedersen.Checked.Digest.choose_preimage h2 in
+            Pedersen.Checked.digest_triples
+              ~init:Hash_prefix.coinbase_merkle_tree.(height)
+              (to_triples h1 @ to_triples h2)
+
+          let assert_equal h1 h2 = Field.Checked.Assert.equal h1 h2
+
+          let if_ = Field.Checked.if_
+        end)
+        (struct
+          include Stack
+
+          type value = t [@@deriving sexp]
+
+          let hash = Checked.digest
+        end)
+
+    let depth = coinbase_tree_depth
+
+    type path = Pedersen.Digest.t list
+
+    module Address = struct
+      include Merkle_tree.Address
+
+      let typ = typ ~depth
+    end
+
+    type _ Request.t +=
+      | Coinbase_stack_path : Address.value -> path Request.t
+      | Get_coinbase_stack : Address.value -> (Stack.t * path) Request.t
+      | Set_coinbase_stack : Address.value * Stack.t -> unit Request.t
+      | Find_index_of_newest_stack : bool -> Address.value Request.t
+      | Find_index_of_oldest_stack : Address.value Request.t
+
+    let reraise_merkle_requests (With {request; respond}) =
+      match request with
+      | Merkle_tree.Get_path addr ->
+          respond (Delegate (Coinbase_stack_path addr))
+      | Merkle_tree.Set (addr, stack) ->
+          respond (Delegate (Set_coinbase_stack (addr, stack)))
+      | Merkle_tree.Get_element addr ->
+          respond (Delegate (Get_coinbase_stack addr))
+      | _ -> unhandled
+
+    let get t addr =
+      handle
+        (Merkle_tree.get_req ~depth (Hash.var_to_hash_packed t) addr)
+        reraise_merkle_requests
+
+    let%snarkydef update_stack t ~is_new_stack stack_before stack_after =
+      let%bind addr =
+        request_witness Address.typ
+          As_prover.(
+            map (read Boolean.typ is_new_stack) ~f:(fun s ->
+                Find_index_of_newest_stack s ))
+      in
+      handle
+        (Merkle_tree.modify_req ~depth (Hash.var_to_hash_packed t) addr
+           ~f:(fun stack ->
+             let%map () =
+               Coinbase_stack.Stack.assert_equal stack_before stack
+             in
+             stack_after ))
+        reraise_merkle_requests
+      >>| Hash.var_of_hash_packed
+
+    let%snarkydef delete_stack t stack_before stack_after =
```

Comment:
This function is essentially the same as the above so I would reuse the code. Also I don't think delete stack is necessarily a good name. What about this function actually involves deleting a stack?